### PR TITLE
[QUIC] Fix native crashes and heap corruption

### LINKED ISSUE #72696: [HTTP/3] SIGABRT in stress tests
Today's stress test run crashed with segmentation fault after 28 mins https://dev.azure.com/dnceng/public/_build/results?buildId=1897576&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=1451f5f3-0108-5a08-5b92-e984b2a85bbd&l=1570

Funny enough, this run was scheduled with this https://github.com/dotnet/runtime/commit/ed5aa3d2a4f9498c0bf2939aab5571565eba2f8b commit on top 😄 @rzikm (might be totally unrelated)

### PR DESCRIPTION
**Pull request by CarnaViire**
There were races between disposing SendBuffers and MsQuic handles and using them from parallel threads. Most prominent in stress tests in configuration `-cancelRate 1 -ops 9` (POST Duplex Dispose).

After the change, I didn't see any crashes in stress in a couple of hours, will keep them running in the meantime.

Change has no impact on perf.

Fixes #72696

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicSafeHandle.cs [hunk 1]
```diff
@@ -52,6 +61,44 @@ protected override bool ReleaseHandle()
...
+        finally
+        {
+            if (success)
+            {
+                DangerousRelease();
+            }
+        }
+    }
+
+    public void SafeCall(Action<MsQuicSafeHandle> call)
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicSafeHandle.cs:83]: This generally should not be need. runtime would automatically grab reference on pinvoke if we pass in safe handle. .... But we don't. It seems like we lost that when switching to generated API. This is all hidden as we pass the safe handle around everywhere and then we don't use it when we should and we pass IntPtr  to quic.

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicSafeHandle.cs:83]: I am not thrilled about the fact that this leads to closure allocation per call in many cases (StreamSend, EnableReceive, ....). I think we should consider (possibly in follow-up) either
- writing wrappers for MsQuicApi which does this for us (i.e. duplicating what P/Invoke does)
- look for a way to have the source generator generate them for us.

Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicSafeHandle.cs:83]: I agree with both comments, and I believe we should look for a way to solve this in generated API, otherwise it is not sustainable. If we would be able to do that in time, it would be ideal.

**Review by rzikm — APPROVED** [review]
LGTM for now, we can follow up with interop improvements in .NET 8.0

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs [hunk 2]
```diff
@@ -392,6 +392,7 @@ public async ValueTask<QuicStream> AcceptInboundStreamAsync(CancellationToken ca
             throw new InvalidOperationException(SR.net_quic_accept_not_allowed);
         }
 
+        GCHandle keepObject = GCHandle.Alloc(this);
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:395]: Why is this here? Isn't this my PR?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs [hunk 3]
```diff
@@ -90,6 +90,7 @@ public sealed partial class QuicStream
         }
     };
     private MsQuicBuffers _sendBuffers = new MsQuicBuffers();
+    private object _sendBuffersLock = new object();
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:93]: :cry:

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs [hunk 4]
```diff
@@ -469,8 +474,8 @@ private unsafe int HandleEventShutdownInitiatedByPeer(ref SHUTDOWN_INITIATED_BY_
     }
     private unsafe int HandleEventShutdownComplete(ref SHUTDOWN_COMPLETE_DATA data)
     {
-        _shutdownTcs.TrySetResult();
         _acceptQueue.Writer.TryComplete(ExceptionDispatchInfo.SetCurrentStackTrace(ThrowHelper.GetOperationAbortedException()));
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:477]: Why the swap?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs (hunk 3 above)
Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:93]: Discussed offline; it's either this or make MsQuicBuffers class instead of struct

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs (hunk 4 above)
Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:477]: Discussed offline; swapping solves the race of ODE vs OperationAbortedException on accepting streams

**Comment by CarnaViire** [comment]
There's a branch with a different approach to safe calls with "generated-like" interop; discussed with @ManickaP that it will be her choice with which solution to proceed
https://github.com/dotnet/runtime/compare/main...CarnaViire:runtime:quic-crash-fix-v2

**Comment by CarnaViire** [comment]
Closing in favor of #74669
